### PR TITLE
Test Restructure for addressing intermittent failures (#419)

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR provides some test reorganization to address some goals:

* We don't want long-running tests to cause build failure on Travis, and sometimes these same temporal tests can fail on slow hardware (e.g. Raspberry PI) or even on local machines in certain circumstances. 
* A bit of cleanup could assure that our tests are getting run and reduce some manual intervention.

Here are some assumptions I discussed on #419 regurgitated:

**Assumption 1**: Some tests are failing on multiple kinds of slow hardware, so I assume that by default slow tests should not get run. That means in practice that the grunt default task would run a subset of all possible tests. It also means Travis would not run the complete set of slow tests (because it's running that default task). Because slow tests are problematic in multiple contexts, it seemed like this would be the appropriate default, but **if this is backwards, it nullifies parts of the approach in this PR.**

**Assumption 2**: All tests in the test directory should be runnable and passing. I found a test, distance.js, that is not included in the list of tests in the nodeunit task on master. I added it back in in this set of changes and it is passing.

**Assumption 3**: `bootstrap.js` test must be run before all test runs of any flavor. There are no other dependencies on test ordering. There are several "core"-ish tests listed at the top of the list in the present `Gruntfile`, but I believe this is just for human organizational sake:

```
        "test/board.js",
        "test/board-connection.js",
        "test/compass.js",
        "test/options.js",
        "test/board.pins.js",
        "test/board.component.js",
        "test/capabilities.js",

```

`Assumption 4`: mock-* scripts are not tests but are supporting mechanisms for running tests.

`Assumption 4a`: It's not weird that lib/board.js references one of the mock- test files (that is presently the case).

### Changes introduced

Introduced a few subdirectories into `test` for organization:

* `bootstrap`: Anything in this directory gets executed on all test runs. Right now that is just `bootstrap.js` and may stay that way but it is an explicit area for essential tests.
* `util`: A directory for utils for tests. These don't get run directly but are required by tests. `mock-*` items are in here.
* `extended`: An area for tests that don't get run in every test run for temporal/long-running reasons, or, potentially later, other reasons. 

I adjusted Grunt tasks:

* `default` and `nodeunit` will run bootstrap test(s) and all tests in `test`. I changed over to using globs for `tests` so that we don't forget to add tests to the task. 
* New task: `nodeunit:complete`. Does everything that `nodeunit` does, *plus* runs extended tests.
* New task: `test`: running `grunt test` is identical to `grunt` (default) but will run the `complete` set instead of the default set. 
* `nodeunit:just` still works and can be used on extended tests, too (documented in a comment).

### Caveats

This PR is *not*:

* A solution to the long-running tests in general.
* A complete separation out of risky, long-running tests into `extended`. For the purposes of illustration, I did pull out the temporal `piezo` tests into `extended`, but there are tests in other modules (e.g. `animation`) that need to be extracted if we use this model.

### CODE REVIEW

**Comment by dtex** [comment]
:relieved: This is awesome! My one concern is that I will forget to run the long running tests when I need to (I'm really good at forgetting things) so I have a selfish request... Could we check for an environment variable in Grunt and run the extended tests if it exists? That way I could set it and forget it.

**Comment by lyzadanger** [comment]
@dtex That is certainly a concern I have as well, which is why I'm highlighting it so much in my assumption list. I'd be down with your environment idea. How would you normally do that? What's your preferred process? Do you know any pattern examples of folks doing this in grunt specifically? Thanks!

File: Gruntfile.js [hunk 1]
```diff
@@ -34,42 +34,8 @@ module.exports = function(grunt) {
...
-        "test/sensor.js",
-        "test/servo.js",
-        "test/shiftregister.js",
-        "test/sonar.js",
-        "test/stepper.js",
-        "test/temperature.js",
-        "test/switch.js",
-        "test/wii.js"
+        "test/bootstrap/*.js",
+        "test/*.js"
```
Comment (rwaldron) [inline Gruntfile.js]: I appreciate what you're trying to do here, but take a look at this discussion: https://github.com/rwaldron/johnny-five/pull/671/files#diff-35b4a816e0441e6a375cd925af50752cR39

**Comment by rwaldron** [comment]
We can land this as soon as [this](https://github.com/rwaldron/johnny-five/pull/679/files#r27422684) is addressed.

File: Gruntfile.js (hunk 1 above)
Comment (lyzadanger) [inline Gruntfile.js]: @rwaldron Let me make sure I understand—it sounds like my assumption 2 is incorrect (that all tests in `test` should be runnable). Would it be acceptable to create a place you could chuck your tests (i.e. another subdir)? I see in the referenced thread that a naming convention isn't gonna fly? Is the only way forward here to go back to listing each test file explicitly?

Comment (rwaldron) [inline Gruntfile.js]: I guess this is really on me to be less lazy about leaving files around. I could get away with it while the list was explicit. Two independent patches seem to be telling me to clean up my act ;)

**Pull request merged**